### PR TITLE
[Dynamic Protocol State] Refactoring to support orthogonal state machine operating on sub-states

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/onflow/flow-go/issues/5556

### Context

This PR implements a huge refactoring for addressing a few outstanding design issues summarized in linked issue. The answer to it was to change the design to one that better fits out needs.

In proposed implementation the design is centered around the KV Store. KV Store is used to store all data related to the Dynamic Protocol State. Processing pipeline was restructured to use orthogonal state machines where they operate on piece of the overall state(a sub-state) and modify only it.

The Epoch state machine was remolded into a hierarchical state machine and it's treated as another orthogonal state machine which is part of the Dynamic Protocol State.

Support for version upgrades/replication has been added as part of this PR as well. 

Conceptually the biggest change is that after this refactoring we are moving from the idea of previous protocol state(`ProtocolStateEntry`, `RichProtocolStateEntry`, etc) will get replaced with KV Store and the previous protocol state becomes part of the KV Store.

:warning: I have disabled large number of tests which all rely on an ability of `Snapshot` to provide KV Store, this is not implemented yet. Tests will be enabled again after implementing: https://github.com/onflow/flow-go/issues/5316

### CODE REVIEW

File: state/protocol/protocol_state.go [hunk 1]
```diff
@@ -117,7 +119,7 @@ type StateMutator interface {
 	//     of the calling code (specifically `FollowerState`).
 	//
 	// updated protocol state entry, state ID and a flag indicating if there were any changes.
-	Build() (hasChanges bool, updatedState *flow.ProtocolStateEntry, stateID flow.Identifier, dbUpdates []transaction.DeferredDBUpdate)
+	Build() (stateID flow.Identifier, dbUpdates []transaction.DeferredDBUpdate, err error)
```
Comment (AlexHentschel) [inline state/protocol/protocol_state.go:122]: I think the gDoc for this method is outdated. Suggestion:

```golang
	// Build constructs the resulting protocol state, *after* applying all the service events via
	// `ApplyServiceEventsFromValidatedSeals(...)`. It returns:
	//  - stateID: the hash commitment to the updated Protocol State Snapshot
	//  - dbUpdates: database updates necessary for (i) persisting the updated protocol state itself including all its
	//    dependencies and (ii) indexing the protocol state as the outcome of this block.
	//  - err: All error returns indicate potential state corruption and should therefore be treated as fatal.
	//
	// CAUTION:
	// For Consensus PARTICIPANTS, the calling code must check that `stateID` matches the commitment in the block proposal!
	Build() (stateID flow.Identifier, dbUpdates []transaction.DeferredDBUpdate, err error)
```

File: state/protocol/protocol_state.go [hunk 2]
```diff
@@ -83,14 +83,16 @@ type ProtocolState interface {
...
+// creating a StateMutator for each block and applying state-changing service events.
 type MutableProtocolState interface {
 	ProtocolState
 
 	// Mutator instantiates a `StateMutator` based on the previous protocol state.
 	// Has to be called for each block to evolve the protocol state.
 	// Expected errors during normal operations:
 	//  * `storage.ErrNotFound` if no protocol state for parent block is known.
-	Mutator(candidateView uint64, parentID flow.Identifier) (StateMutator, error)
+	Mutator(candidate *flow.Header) (StateMutator, error)
```
Comment (AlexHentschel) [inline state/protocol/protocol_state.go:95]: ### ⚠️  

I strongly disagree with this change. There are two distinct scenarios that the Mutator has to satisfy:
1. Checking another node's block proposal. Here, we have a candidate block. All good. 
2. Constructing our own block proposal: 

   Here, we do **_not_ have a candidate header**! The header contains the payload's hash and the payload contains the commitment to the output state. We instantiate the `Mutator` to compute the output state. Therefore, by causality the primary cannot have a valid block header. 

In conclusion, a proper `flow.Header` does not exist for all application scenarios of this method.

Comment (AlexHentschel) [reply state/protocol/protocol_state.go:95]: _Further Context:_ 

I am aware that the current way of building our blocks has this notion of "header under construction" and the type of this header is unfortunately also `flow.Header`.  I consider this _very_ bad technical debt that urgently needs to be resolved (though not part of this work stream). There are a bunch of really nasty complexities resulting from this tech debt:
* You can ask a "header under construction" for its ID. It will give you one, but that is not the ID of the final block, because some fields have dummy values.  This is an invitation for subtle bugs. If I am not mistaken, you stepped right into the trap already in your own PR [here](https://github.com/onflow/flow-go/blob/d69920a457a1dbde4c2a01485414d8645a1b810f/state/protocol/protocol_state/state/mutator.go#L106-L108).
* On a broader scale of the code base, exactly this non-existent differentiation between `entity under construction` and the `final entity` is what is causing all the malleability edge-cases in our code base: 
   * Engineer A adds field `x` to the entity but explicitly excludes it from the entity's ID computation, because the entity is still under construction, and he wants to refer to the entity with the same ID despite still needing to add that one last field. 
   * Engineer B (or Engineer A after some months) sees the field `x` in the entity and uses it for something else. Though, as `x` is not covered by the hash computation, byzantine actors can manipulate the value at will and continuously. Furthermore, they can use different values of `x` to undermine the networking layer's deduplication logic to spam the application layer. 

Therefore, we will (hopefully this year) add a clean separation by types for "entity under construction" vs "final entity". For the primary, you will simply only have the "header under construction" when initializing the `Mutator` which cannot give you a blockID to index the protocol state with!

Comment (AlexHentschel) [reply state/protocol/protocol_state.go:95]: Suggestion: 
 * keep the previous interface
 * you cannot move the indexing logic into the Mutator for causality reasons

File: state/protocol/badger/mutator.go [hunk 3]
```diff
@@ -608,18 +608,14 @@ func (m *FollowerState) evolveProtocolState(ctx context.Context, candidate *flow
...
+	if err != nil {
+		return fmt.Errorf("could not build dynamic protocol state: %w", err)
+	}
 	if updatedStateID != candidate.Payload.ProtocolStateID {
 		return state.NewInvalidExtensionErrorf("invalid protocol state commitment %x in block, which should be %x", candidate.Payload.ProtocolStateID, updatedStateID)
 	}
-
-	// Schedule deferred database operations to index the protocol state by the candidate block's ID
-	// and persist the new protocol state (if there are any changes)
-	deferredDbOps.AddDbOp(m.protocolStateSnapshotsDB.Index(candidate.ID(), updatedStateID))
```
Comment (AlexHentschel) [inline state/protocol/badger/mutator.go:618]: ## ⚠️ 

you cannot move this logic into the `Mutator` for causality reasons as explained [here](url)

**Review by AlexHentschel — CHANGES_REQUESTED** [review]
I think we have to reverse the API changes, where the `Mutator` is not initialized with the candidate header as opposed to the view and parent. See [my comment here](url) for details. Still lots to review for me but wanted to flag this concern asap.

File: state/protocol/protocol_state/state/mutator.go [hunk 4]
```diff
@@ -100,17 +92,33 @@ func newStateMutator(
...
+	}
+	stateID := m.kvMutator.ID()
+	version, data, err := m.kvMutator.VersionedEncode()
+	if err != nil {
+		return flow.ZeroID, nil, fmt.Errorf("could not encode protocol state: %w", err)
+	}
+
+	// Schedule deferred database operations to index the protocol state by the candidate block's ID
+	// and persist the new protocol state (if there are any changes)
+	dbUpdates = append(dbUpdates, m.kvStoreSnapshots.IndexTx(m.candidate.ID(), stateID))
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/state/mutator.go:108]: ⚠️ I think this is wrong, because when the primary executed this logic, the `candidate` header cannot give you a valid `ID` as explained in [my comment above](https://github.com/onflow/flow-go/pull/5616#discussion_r1548696603). We are tripping over our own malleability foot-guns 😱.

File: state/protocol/protocol_state/epochs/statemachine.go [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+// implementations, whose constructors both have the same signature as StateMachineFactoryMethod.
+type StateMachineFactoryMethod func(candidateView uint64, parentState *flow.RichProtocolStateEntry) (StateMachine, error)
+
+// EpochStateMachine is a hierarchical state machine that encapsulates the logic for protocol-compliant evolution of Epoch-related sub-state.
+// EpochStateMachine processes a subset of service events that are relevant for the Epoch state, and ignores all other events.
+// EpochStateMachine delegates the processing of service events to an embedded StateMachine,
+// which is either a HappyPathStateMachine or a FallbackStateMachine depending on the operation mode of the protocol.
+// It relies on Key-Value Store to commit read the parent state and to commit the updates to the Dynamic Protocol State.
+type EpochStateMachine struct {
+	candidate                        *flow.Header
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/statemachine.go:85]: ⚠️ causal circle. Header of the output block is not available while the primary builds the block initially. 

Specifically, this line https://github.com/onflow/flow-go/blob/d69920a457a1dbde4c2a01485414d8645a1b810f/state/protocol/protocol_state/epochs/statemachine.go#L170 is not generally correct in the context we execute this logic

File: state/protocol/protocol_state/state/mutator.go [hunk 6]
```diff
@@ -100,17 +92,33 @@ func newStateMutator(
...
+func (m *stateMutator) Build() (flow.Identifier, []transaction.DeferredDBUpdate, error) {
+	var dbUpdates []transaction.DeferredDBUpdate
+	for _, stateMachine := range m.orthoKVStoreMachines {
+		dbUpdates = append(dbUpdates, stateMachine.Build()...)
+	}
+	stateID := m.kvMutator.ID()
+	version, data, err := m.kvMutator.VersionedEncode()
+	if err != nil {
+		return flow.ZeroID, nil, fmt.Errorf("could not encode protocol state: %w", err)
+	}
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/state/mutator.go:104]: 😭 I think I understand now why you implemented it that way:
* not only do we want to index the KV-Store by the ID of the resulting block
* but we also want that individual sub-state machines can maintain their own `ByBlock` indices. That makes sense and is a very strong feature. 

For this purpose, I think `transaction.DeferredDBUpdate` is too simple:
* `transaction.DeferredDBUpdate` assumes that all information is already available at the time we construct the pending DB updates.
* However, this does not apply to the ID of the resulting block. 

I think we want something more powerful. A construct where we can say: "Once we are done with the construction and have the final block ID, _then_ we want to apply these updates and indexing operations"

I would suggest to structure this as a _wrapper_ around `transaction.DeferredDBUpdate`

**Review by AlexHentschel — COMMENTED** [review]
The scope of the problem is getting more clear now. Unfortunately, it is bigger than I had hoped. See specifically [this comment](https://github.com/onflow/flow-go/pull/5616#discussion_r1548934955)

File: state/protocol/protocol_state.go [hunk 7]
```diff
@@ -117,7 +119,7 @@ type StateMutator interface {
 	//     of the calling code (specifically `FollowerState`).
 	//
 	// updated protocol state entry, state ID and a flag indicating if there were any changes.
-	Build() (hasChanges bool, updatedState *flow.ProtocolStateEntry, stateID flow.Identifier, dbUpdates []transaction.DeferredDBUpdate)
+	Build() (stateID flow.Identifier, dbUpdates DeferredDBUpdates, err error)
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state.go:122]: +1, see [my suggestion above](https://github.com/onflow/flow-go/pull/5616#discussion_r1548688184)

File: state/protocol/protocol_state.go [hunk 8]
```diff
@@ -83,14 +83,16 @@ type ProtocolState interface {
...
+// creating a StateMutator for each block and applying state-changing service events.
 type MutableProtocolState interface {
 	ProtocolState
 
 	// Mutator instantiates a `StateMutator` based on the previous protocol state.
 	// Has to be called for each block to evolve the protocol state.
 	// Expected errors during normal operations:
 	//  * `storage.ErrNotFound` if no protocol state for parent block is known.
-	Mutator(candidateView uint64, parentID flow.Identifier) (StateMutator, error)
+	Mutator(view uint64, parentID flow.Identifier) (StateMutator, error)
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state.go:95]: +1

File: state/protocol/protocol_state/epochs/statemachine.go [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+	"github.com/onflow/flow-go/storage/badger/transaction"
+)
+
+// StateMachine implements a low-level interface for state-changing operations on the Epoch state.
+// It is used by higher level logic to coordinate the Epoch handover, evolving its internal state
+// when Epoch-related Service Events are sealed or specific view-thresholds are reached.
+//
+// The StateMachine is fork-aware, in that it starts with the Epoch state of the parent block and
+// consumes a sequence of sealed Service Events from the child block. A separate instance is created for
+// each block that is being processed. Calling `Build()` constructs a snapshot of the resulting Epoch state.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/epochs/statemachine.go:21]: I think we could be a bit more concrete: sealed service events and view is literally the only information that the Epoch State Machine processes. Suggestion:

```golang
⋮
//
// The StateMachine is fork-aware, in that it starts with the Epoch state of the parent block and
// evolves the state based on the relevant information in the child block (specifically Service Events
// sealed in the child block and the child block's view). A separate instance must be created for each
// block that is being processed. Calling `Build()` constructs a snapshot of the resulting Epoch state.
type StateMachine interface {
```

File: state/protocol/protocol_state/state/mutator.go [hunk 10]
```diff
@@ -100,17 +91,35 @@ func newStateMutator(
 //     of the calling code (specifically `FollowerState`).
 //
 // updated protocol state entry, state ID and a flag indicating if there were any changes.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/state/mutator.go:93]: suggesting [this goDoc](https://github.com/onflow/flow-go/pull/5616#discussion_r1550619778)

File: state/protocol/protocol_state/epochs/statemachine.go [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+	})
+	if hasChanges {
+		dbUpdates.AddBadgerUpdate(operation.SkipDuplicatesTx(e.protocolStateDB.StoreTx(updatedStateID, updatedEpochState)))
+	}
+	e.mutator.SetEpochStateID(updatedStateID)
+
+	return dbUpdates
+}
+
+// ProcessUpdate applies the state changes that are delivered via sealed service events.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/epochs/statemachine.go:184]: > this function must be invoked even if there are no "updates" to pass in

Thanks Jordan. I think this is a really important concern. My preference would be to mediate this risk through a callout in the goDoc and maybe renaming the function. I think that would be sufficient.

File: state/protocol/protocol_state/epochs/base_statemachine.go [hunk 12]
```diff
@@ -83,30 +24,30 @@ type baseProtocolStateMachine struct {
 
 // Build returns updated protocol state entry, state ID and a flag indicating if there were any changes.
 // CAUTION:
-// Do NOT call Build, if the ProtocolStateMachine instance has returned a `protocol.InvalidServiceEventError`
-// at any time during its lifetime. After this error, the ProtocolStateMachine is left with a potentially
+// Do NOT call Build, if the StateMachine instance has returned a `protocol.InvalidServiceEventError`
+// at any time during its lifetime. After this error, the StateMachine is left with a potentially
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/epochs/base_statemachine.go:28]: > this function must be invoked even if there are no "updates" to pass in

Thanks Jordan. I think this is a really important concern. My preference would be to mediate this risk through better documentation and maybe renaming the function. I think that would be sufficient: 

* If we **specifically mention view-based triggers**, I think that will help the readers understand this requirement. 
* In addition, the `CAUTION` statement you are suggesting would help.

File: state/protocol/protocol_state/epochs/statemachine.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/epochs/statemachine.go:184]: suggestion

```golang
// EvolveState evolves the Epoch state in accordance with the information in the candidate
// block (under construction). Information that potentially changes the Epoch state is:
//   - Service Events sealed in the candidate block
//   - the candidate block's view (already provided at construction time)
//
// CAUTION: This function MUST be called for all candidate blocks, even if `update` is empty! This
// is because also the absence of expected service events by a certain view can result in the Epoch
// state changing (for example not having received the EpochCommit event for the next epoch, but
// approaching the end of the current epoch).
//
// The block's payload might contain epoch preparation service events for the next
// epoch. In this case, we need to update the tentative protocol state.
// We need to validate whether all information is available in the protocol
// state to go to the next epoch when needed. In cases where there is a bug
// in the smart contract, it could be that this happens too late, and we should trigger epoch fallback mode.
// No errors are expected during normal operations.
func (e *EpochStateMachine) EvolveState(sealedServiceEvents []flow.ServiceEvent) error {
```

besides the extended goDoc, I am suggesting the following refinements:
* rename the input `update` to `sealedServiceEvents` for clarity 
* suggesting to rename the function to `EvolveState`. I somewhat feel that this naming provides a clearer separation between the inputs and what the method does (previously we used "update" both in the function name as well as the input, which I think is part of the concern here).

File: state/protocol/protocol_state/kvstore.go [hunk 13]
```diff
@@ -105,25 +116,76 @@ type KVStoreMutator interface {
...
+	// Build returns:
+	//   - database updates necessary for persisting the updated protocol sub-state and its *dependencies*.
+	//     It may contain updates for the sub-state itself and for any dependency that is affected by the update.
+	//     Deferred updates must be applied in a transaction to ensure atomicity.
+	Build() protocol.DeferredDBUpdates
+
+	// ProcessUpdate processes an ordered list of sealed service events.
+	// Usually, each state machine performs filtering of relevant events and ignores all other events.
+	// No errors are expected during normal operations.
+	ProcessUpdate(orderedUpdates []flow.ServiceEvent) error
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/kvstore.go:170]: this is an iteration on [Jordan's suggestion here](https://github.com/onflow/flow-go/pull/5616#discussion_r1549827439): 

```suggestion
	// EvolveState applies the state change(s) on sub-state P for the candidate block (under construction).
	// Information that potentially changes the Epoch state (compared to the parent block's state):
	//   - Service Events sealed in the candidate block
	//   - the candidate block's view (already provided at construction time)
	//
	// CAUTION: EvolveState MUST be called for all candidate blocks, even if `sealedServiceEvents` is empty!
	// This is because also the absence of expected service events by a certain view can also result in the
	// Epoch state changing (for example not having received the EpochCommit event for the next epoch, but
	// approaching the end of the current epoch).
	//
	// No errors are expected during normal operations.
	EvolveState(sealedServiceEvents []flow.ServiceEvent) error
```

File: state/protocol/protocol_state/epochs/statemachine.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/epochs/statemachine.go:184]: _Change of opinion_ 😅 

I think the patter of "this function must be called even if there is no input" is a vastly unsafe pattern, because it creates this additional edge case across multiple abstraction layers. I think we would need `CAUTION` statements at [this layer ](https://github.com/onflow/flow-go/blob/63d0114b1335b16d561854dfd027fc6d66bc8d2b/state/protocol/protocol_state.go#L180)for example. I think we can simplify the usage patterns of our API in this regards. I another words, I think we need to fix it (`TODO` - not in this PR)

File: state/protocol/protocol_state.go [hunk 14]
```diff
@@ -177,3 +179,37 @@ type StateMutator interface {
 	//     of this function is not nil. If such an exception is returned, continuing is not an option.
 	ApplyServiceEventsFromValidatedSeals(seals []*flow.Seal) error
 }
+
+type DeferredDBUpdate func(tx *transaction.Tx, blockID flow.Identifier) error
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state.go:183]: I would suggest to call this struct `DeferredBlockPersistOp`, because we are literally letting the state machines specify the db operations to persist their respective the sub-state. All that persisted data is strictly tied to a block, which I think is important to reflect in the code.

**Review by jordanschalm — APPROVED** [review]
Got through the remaining files. A few minor suggestions, nothing major though. Great work!

File: state/protocol/protocol_state/kvstore/upgrade_statemachine.go [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+	// activation view is higher than the sealing view + Δ when accepting the event. Δ represents the finalization lag
+	// to give time for replicas to finalize the block containing the seal for the version upgrade event.
+	// When replica reaches activation view and the latest finalized protocol state knows about the version upgrade,
+	// then it's safe to switch the protocol version.
+	if m.view+m.params.EpochCommitSafetyThreshold() >= versionUpgrade.ActiveView {
+		return protocol.NewInvalidServiceEventErrorf("invalid protocol state version upgrade view %d -> %d: %w",
+			m.view+m.params.EpochCommitSafetyThreshold(), versionUpgrade.ActiveView, ErrInvalidActivationView)
+	}
+
+	if m.parentState.GetProtocolStateVersion() >= versionUpgrade.NewProtocolStateVersion {
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/kvstore/upgrade_statemachine.go:97]: +1 on only permitting increments by 1

File: state/protocol/protocol_state/kvstore/upgrade_statemachine.go [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+	// check in case there is a pending upgrade in parent state.
+	err := checkPendingUpgrade(m.parentState)
+	if err != nil {
+		return fmt.Errorf("version upgrade invalid with respect to the parent state: %w", err)
+	}
+	// check in case there are multiple upgrades in the same block.
+	err = checkPendingUpgrade(m.mutator)
+	if err != nil {
+		return fmt.Errorf("version upgrade invalid with respect to the current state: %w", err)
+	}
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/kvstore/upgrade_statemachine.go:133]: quite understand why we have to check the parent state. Aren't these all transitive checks? In other words, if I already have checked the compatibility of Activator1 ->  Activator2 -> ... -> ActivatorK and now I am finding that it is legal to modify ActivatorK -> ActivatorL. Then by transitivity, wouldn't it hold that I could directly go from Activator1 -> ActivatorL?

In other words, if we encounter multiple service events, we would write the newest one into `m.mutator` and if we haven't encountered any Service events, `m.mutator` contains the latest activator from the parent. In all cases, we only need to compare to `m.mutator` I think. 
 
```suggestion
	// There can be multiple `versionUpgrade` Service Events sealed in one block. In case we have _not_
	// encountered any, `m.mutator` contains the latest `versionUpgrade` as of the parent block, because
	// we cloned it from the parent state. If we encountered some version upgrades, we already enforced that
	// they are all upgrades to the same version. So we only need to check that the next `versionUpgrade`
	// also has the same version.
	err := checkPendingUpgrade(m.mutator)
	if err != nil {
		return fmt.Errorf("version upgrade invalid with respect to the current state: %w", err)
	}
```

File: state/protocol/protocol_state/state/mutator_test.go
Comment (AlexHentschel) [inline state/protocol/protocol_state/state/mutator_test.go:1]: great tests!

**Review by AlexHentschel — APPROVED** [review]
Great PR (and huge). Only minor suggestions. I have some thoughts on how to make `DeferredDBUpdate` a bit prettier ... but lets get this PR merged first.

File: state/protocol/protocol_state/kvstore/upgrade_statemachine.go (hunk 16 above)
Comment (durkmurder) [reply state/protocol/protocol_state/kvstore/upgrade_statemachine.go:136]: Yeah, if we consider the fact that replicated state is a copy of parent state then sure, we need to check only the current state. I haven't thought about this, applying change.

File: state/protocol/protocol_state/kvstore/models.go [hunk 17]
```diff
@@ -9,22 +9,31 @@ import (
...
+// for a particular Protocol State version.
+// Essentially enumerating the set of keys and values that are supported.
+// When a key is added or removed, this requires a new protocol state version.
+// To use new version of the protocol state, create a new versioned model in models.go (eg. modelv3 if latest model is modelv2)
+// ATTENTION: All models should be public with public fields otherwise the encoding/decoding will not work.
+
+// UpgradableModel is a utility struct that can be embedded in any model to provide
+// a common interface for managing protocol version upgrades.
+type UpgradableModel struct {
 	VersionUpgrade *protocol_state.ViewBasedActivator[uint64]
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply state/protocol/protocol_state/kvstore/models.go:22]: not sure about this, the whole reason for it to be a pointer is to be an optional value. We don't always have `VersionUpgrade` available, one example is spinning up a chain so a pointer made a lot of sense to me, especially to avoid introducing some initial state to the `ViewBasedActivator[uint64]`

File: state/protocol/protocol_state/epochs/statemachine_test.go [hunk 18]
```diff
@@ -0,0 +1,512 @@
...
+// epoch state ID in the KV store even when there were no events to process.
+func (s *EpochStateMachineSuite) TestBuild_NoChanges() {
+	s.happyPathStateMachine.On("ParentState").Return(s.parentEpochState)
+	s.happyPathStateMachine.On("Build").Return(s.parentEpochState.ProtocolStateEntry, s.parentEpochState.ID(), false).Once()
+
+	err := s.stateMachine.ProcessUpdate(nil)
+	require.NoError(s.T(), err)
+
+	indexTxDeferredUpdate := storagemock.NewDeferredDBUpdate(s.T())
+	indexTxDeferredUpdate.On("Execute", mocks.Anything).Return(nil).Once()
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply state/protocol/protocol_state/epochs/statemachine_test.go:102]: the `Execute` is a general mocked method for all anonymous functions. Since a `DeferredDBUpdate` is a anonymous function we expect call to `Execute` since we pass it in next [line](https://github.com/onflow/flow-go/pull/5616/files/871174eb00fdc1149bbc1c7bafe94f43fa5a6e4c#diff-f20f81e74ce9a22d009f7300d0a42cd62faa7885cfd78ae05b8928ead315f506R104).

File: state/protocol/protocol_state/epochs/statemachine.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (durkmurder) [reply state/protocol/protocol_state/epochs/statemachine.go]: I think we are getting way down to a rabbit hole with this reasoning. We are as engineers are responsible for calling functions in correct order with correct arguments. To me there is little value is adding extra constructions to ensure that `EvolveState` will be called, there is exactly one place where it is called, it's part of encapsulated logic, it's not accessible to client or user in any way. To me it seems that we are trying to invent a mechanism that will enforce that `EvolveState` is called for each state machine, but that is unnecessary since the calling code is very straight-forward, it's literally a for loop which is initiated as part of the pipeline that _we_ have built. Happy to continue this discussion.

**Comment by durkmurder** [comment]
@AlexHentschel @jordanschalm FYI, I have commented out a much needed check for integrity of the sealing segment(which I have previously added as suggest from Alex) because we cannot support it in current state, due to different protocol state IDs at root block and the very next one.
